### PR TITLE
feat(cart): transform cart errors with injectable error matcher

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Cart errors are stored as `string`s.

Fixes: #1107 

## What is the new behavior?
Cart errors are stored as `DaffStateError`s. Effects transform the errors coming out of the driver layer with the injected error matcher function.

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

Changes the payload type of failure actions from `string` to `DaffStateError`.

## Other information
I'm sorry that this PR is so large. LMK if you want me to break it up into feature slices.

### CODE REVIEW

File: libs/cart/state/src/effects/cart-resolver.effects.ts [hunk 1]
```diff
@@ -4,8 +4,9 @@ import { Action } from '@ngrx/store';
 import { Observable, of } from 'rxjs';
 import { switchMap, catchError, map, mapTo } from 'rxjs/operators';
 
-import { DaffStorageServiceError } from '@daffodil/core';
-import { DaffCart, DaffCartStorageService } from '@daffodil/cart';
+import { DaffStorageServiceError, DaffError } from '@daffodil/core';
+import { substream } from '@daffodil/core/state';
```
Comment (lderrickable) [inline libs/cart/state/src/effects/cart-resolver.effects.ts:8]: unused import

File: libs/cart/state/src/effects/cart-resolver.effects.spec.ts [hunk 2]
```diff
@@ -3,12 +3,14 @@ import { provideMockActions } from '@ngrx/effects/testing';
 import { Observable, of } from 'rxjs';
 import { hot, cold } from 'jasmine-marbles';
 
-import { DaffStorageServiceError } from '@daffodil/core';
+import { DaffError, DaffStorageServiceError } from '@daffodil/core';
```
Comment (lderrickable) [inline libs/cart/state/src/effects/cart-resolver.effects.spec.ts:6]: unused import

**Review by lderrickable — CHANGES_REQUESTED** [review]

File: libs/cart/state/src/effects/cart-address.effects.ts [hunk 3]
```diff
@@ -4,8 +4,9 @@ import { of } from 'rxjs';
 import { Actions, Effect, ofType } from '@ngrx/effects';
 
 import { DaffStorageServiceError } from '@daffodil/core';
-import { DaffCartAddress, DaffCart, DaffCartStorageService } from '@daffodil/cart';
+import { DaffCartAddress, DaffCart, DaffCartStorageService, DAFF_CART_ERROR_MATCHER } from '@daffodil/cart';
 import { DaffCartAddressDriver, DaffCartAddressServiceInterface } from '@daffodil/cart/driver';
+import { daffTransformErrorToStateError } from '@daffodil/core/state';
```
[the remark this answers is not in the document]
Comment (griest024) [reply libs/cart/state/src/effects/cart-address.effects.ts]: removed

File: libs/cart/state/src/effects/cart-resolver.effects.ts (hunk 1 above)
Comment (griest024) [reply libs/cart/state/src/effects/cart-resolver.effects.ts]: removed

File: libs/cart/state/src/effects/cart-resolver.effects.spec.ts (hunk 2 above)
Comment (griest024) [reply libs/cart/state/src/effects/cart-resolver.effects.spec.ts]: removed

**Review by lderrickable — APPROVED** [review]

**Comment by lderrickable** [comment]
lgtm

**Pull request merged**